packages/connect-core/src/transform-iterable-story.spec.ts

```diff
@@ -33,6 +33,78 @@ import {
   transformParse,
 } from "./transform-iterable.js";
 
+class Reader<T> {
+  private it: AsyncIterable<T>;
+
+  constructor(it: AsyncIterable<T>) {
+    this.it = it;
+  }
+  [Symbol.asyncIterator](): AsyncIterator<T> {
+    return {
+      next: async () => {
+        for await (const item of this.it) {
+          return { value: item };
+        }
+        return { value: "", done: true };
+      },
+    };
+  }
+}
+
+class Writer<T> {
+  private queue: IteratorResult<T>[] = [];
+  // Represents the resolve function of the promise returned by the async iterator if no values exist in the queue at
+  // the time of request.  It is resolved when a value is successfully received into the queue.
+  private queueResolver: ((val: IteratorResult<T>) => void) | undefined;
+  // Represents the resolve function of the promise returned by send.  It is resolved when a value is successfully
+  // read via the async iterator.
+  private sendResolver: (() => void) | undefined;
+
+  async process(item: IteratorResult<T, undefined>) {
+    // If there is an iterator resolver then a consumer of the async iterator is waiting on a value.  So resolve that
+    // promise with the new value being sent and return a promise that is immediately resolved
+    if (this.queueResolver) {
+      this.queueResolver(item);
+      this.queueResolver = undefined;
+      return new Promise<void>((resolve) => {
+        resolve();
+      });
+    }
+    // Otherwise no one is waiting on a value yet so add it to the queue and return a promise that will be resolved
+    // when someone reads this value
+    this.queue.push(item);
+    return new Promise<void>((resolve) => {
+      this.sendResolver = resolve;
+    });
```

Comment:
Consider the case where the user calls send() multiple times, without waiting for the returned promise to resolve:

```ts
function f(w: Writer<string>): Promise {
  return Promise.all(
    w.send("a"),
    w.send("b"),
    w.send("c"),
  );
}
```

In this case, the promises for "a" and "b" are never resolved, because `sendResolver` is overwritten by sending "c".

I think we have two options here: 1. Prevent concurrent send by raising an error. 2. Push not just the value to be sent onto the queue, but also its `sendResolver`. I think option 2 is more forgiving.

I think we should add a test for this situation.